src/python/pants/core/goals/clean.py

```diff
@@ -0,0 +1,158 @@
+from pants.core.util_rules.distdir import DistDir
+from pants.engine.console import Console
+from pants.engine.goal import Goal, GoalSubsystem
+from pants.engine.internals.options_parsing import _Options
+from pants.engine.rules import collect_rules, goal_rule
+from pants.option.global_options import LocalStoreOptions
+from pants.util.dirutil import rm_rf
+
+# N.B. We use a list so that named-caches can go last, as it is quite large and for dry-run
+# it's a nicer UX if it's last.
+_NAMED_CACHE_FLAGS_INFO = [
+    (
+        "process-execution-temp-dirs",
+        "Whether to keep the process execution temporary directories. These are normally cleaned up"
+        " automatically, however they are leaked if --no-process-execution-local-cleanup.",
+    ),
+    (
+        "lmdb-store",
+        "Whether to keep the local file store, which stores the results of subprocesses run by Pants.",
+    ),
+    (
+        "run-tracker",
+        "Whether to clean directories persisted after using `./pants run`.",
+    ),
+    (
+        "repl-temp-dirs",
+        "Whether to clean directories persisted after using `./pants repl`.",
+    ),
+    ("dist-dir", "Whether to clean the distributable files directory. "),
+    ("bootstrap-dir", "Whether to clean directories used by pants for bootstrapping itself."),
+    ("named-caches", "Whether to clean the named global caches."),
+]
+
+
+def _dir_info(*paths: Path):
+    total_size = num_files = 0
+    for path in paths:
+        for dirpath, _, filenames in os.walk(path):
+            total_size += sum(os.path.getsize(os.path.join(dirpath, f)) for f in filenames)
+            num_files += len(filenames)
+    return total_size, num_files
+
+
+class CleanSubsystem(GoalSubsystem):
+    name = "clean"
+    help = "Clean pants caches and temp dirs (based on current settings)"
+
+    @classmethod
+    def register_options(cls, register) -> None:
+        super().register_options(register)
+        register(
+            "--dry-run",
+            advanced=False,
+            type=bool,
+            default=True,
+            help=("Whether to just report what would be cleaned. @TODO: more"),
+        )
+        register(
+            "--all",
+            advanced=False,
+            type=bool,
+            default=False,  # @TODO: True?
+            help=(
+                "Clear all caches. Can be combined with the advanced options for further control."
+            ),
+        )
+
+        # @TODO: Is there a way to collect these programmatically?
+        for flag_slug, help in _NAMED_CACHE_FLAGS_INFO:
+            register(
+                f"--keep-{flag_slug}",
+                advanced=True,
+                type=bool,
+                default=True,
+                help=help,
+            )
+
+
+class Clean(Goal):
+    subsystem_cls = CleanSubsystem
+
+
+@goal_rule
+async def clean(
+    clean_subsystem: CleanSubsystem,
+    console: Console,
+    real_opts: _Options,
```

Comment:
You can request `GlobalOptions` instead

Reply:
What about getting to the `store_dir`? `LocalStoreOptions` doesn't seem to jive as a parameter :thinking:

Reply:
Wanted to make sure that you had seen https://github.com/pantsbuild/pants/issues/11167 : ... part of the reason it suggests that this should be a "builtin" goal is that interacting with the store directory while `pants` is running has the potential to cause crashes. And there are native methods that allow for garbage collecting to a target size, etc.

If you're willing to back up and take that tack instead, I think that it will allow for some more subtlety here? If not, I think that we should consider landing this as `./pants experimental-clean` so that we have some more flexibility to change the default modes and options.

Reply:
Oh no haven't seen that one. Let me sit on it a while and do some thinking. I definitely like the idea, but need some time to do mental mapping.

Reply:
Based on [my comment](https://github.com/pantsbuild/pants/issues/11167#issuecomment-962462392) I can change this to be `./pants caches report` and `./pants caches clean`. I'll need some help on how to instrument sub-commands (sub-goals?).